### ISSUE TITLE
Vulkan support in MacOS?

### ISSUE THREAD

**Issue by palmerj** [RobertBeckebans/RBDOOM-3-BFG#763]
I just tried compiling rboom3 with Vulkan support on MacOS M2Max. It's builds and is functional. However when I play the game the framerate is really low and textures are not displaying correctly (see screenshot below). Is Vulkan support on MacOS ready for prime time?

<img width="1392" alt="Screenshot 2023-04-18 at 4 20 25 pm" src="https://user-images.githubusercontent.com/580022/232689311-b4923669-adc1-4f59-b9be-1773d2978daa.png">

**Comment by asemarafa** [comment]
Same performance issue on m1 pro, i moved to version 1.4 instead for opengl

**Comment by SRSaunders** [comment]
I have been working on this problem in the background and believe it is now solved by PR #818.  Please retest after applying this pull request to the master branch and let me know if it fixes your issues.

**Comment by asemarafa** [comment]
There is a notable improvement in FPS with Vulkan after the patch, especially during aircraft takeoff where FPS previously dropped below 10. The performance is also significantly better when compared to OpenGL.

#### Vulkan Unpatched
![Vulkan Unpatched](https://github.com/RobertBeckebans/RBDOOM-3-BFG/assets/30239512/dd25134f-a3a3-4099-894d-80c39b94b9c7)

#### OpenGL
![OpenGL](https://github.com/RobertBeckebans/RBDOOM-3-BFG/assets/30239512/bf45c56d-ab42-40d1-8aaa-b7415b0a9b92)

#### Vulkan Patched
![Vulkan Patched](https://github.com/RobertBeckebans/RBDOOM-3-BFG/assets/30239512/e248ecf0-a1ea-45fd-b489-ce92703095fa)
---

**Comment by SRSaunders** [comment]
This is all excellent news.  Thank you very much for testing and demonstrating the improvement using the new Vulkan nvrhi stream.  The only odd thing is the messed-up GPU-side timing info in the FPS overlay.

Which version of the Vulkan SDK are you using?  If not the latest, would you mind installing the 1.3.261.1 version from LunarG (containing MoltenVK 1.2.5) and retesting the last case (Vulkan Patched) above?  I believe this should solve the GPU-side timing stats issues for Apple Silicon machines.  Please let me know.

**Comment by SRSaunders** [comment]
FYI I have raised an issue https://github.com/KhronosGroup/MoltenVK/issues/2039 with MoltenVK regarding GPU timestamp calibration in `vulkansdk-macos-1.3.261.1`

On a different topic, I would like to ask about your experience using the pre-canned `cmake-xcode-xxx.sh` files on Apple Silicon when configuring the RBDoom3BFG build.  Are you using `brew` or `MacPorts`?  And did you have to edit the shell script to alter any build instructions?  I am considering updating the shell scripts for macOS/Xcode to be more friendly to Apple Silicon since that is becoming the dominant platform.  I would appreciate your perspective as a user.

**Comment by asemarafa** [comment]
I've modified the cmake-macos-release.sh release script to work on both my Intel and M1 Macs without requiring Xcode. My setup relies on Homebrew and aims for quick command-line compilation.

I've added a variable for the openal-soft prefix to make it more flexible. Here's the snippet for your reference:

```
cd ..
rm -rf build
mkdir build
cd build

# Determine the Homebrew prefix for openal-soft
OPENAL_PREFIX=$(brew --prefix openal-soft 2>/dev/null)

if [ -z "$OPENAL_PREFIX" ]; then
  echo "Error: openal-soft is not installed via Homebrew."
  echo "Either install it using 'brew install openal-soft' or provide the path manually."
  exit 1
fi

# change or remove -DCMAKE_OSX_DEPLOYMENT_TARGET=<version> to match supported runtime targets
cmake -G "Unix Makefiles" -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_FLAGS_RELEASE="-DNDEBUG" -DCMAKE_OSX_DEPLOYMENT_TARGET=12.1 -DFFMPEG=OFF -DBINKDEC=ON -DUSE_MoltenVK=ON -DOPENAL_LIBRARY=$OPENAL_PREFIX/lib/libopenal.dylib -DOPENAL_INCLUDE_DIR=$OPENAL_PREFIX/include ../neo -Wno-dev

```

I noticed that the README assumes users are going to utilize Xcode for Apple Silicon builds. Given my changes, should I submit a PR?

**Comment by SRSaunders** [comment]
This looks good, but I want to fix all the single architecture build scripts (i.e. non-universal) to support both x86 and Apple Silicon using Homebrew.  This means adding your changes to:

```
cmake-macos-release.sh
cmake-macos-retail.sh
cmake-xcode-debug.sh
cmake-xcode-release.sh
```

I would also like to fix the `cmake-xcode-universal.sh` script to test for MacPorts and make sure the openal-soft universal variant is installed.

Would you mind if I added your changes as I have described here?  I would add a new commit to my open PR and credit you for your code.

**Comment by asemarafa** [comment]
Sure, go ahead.

**Comment by SRSaunders** [comment]
Thanks.  Your changes added to #818.

**Comment by SRSaunders** [comment]
FYI @asemarafa - Issue https://github.com/KhronosGroup/MoltenVK/issues/2039 regarding GPU timers on Apple Silicon is fixed in MoltenVK 1.2.6.  This will be shipped as part of the upcoming Vulkan SDK 1.3.268.  Once this is available the RBDoom3BFG HUD display will again work correctly for GPU timers on Apple Silicon.
